app/lib/package/export_api_to_cdn.dart

```diff
@@ -0,0 +1,239 @@
+    try {
+      await lock.withClaim((claim) async {
+        await doUploadPackageApisToCdnBucket(claim, bucket);
+      });
+    } catch (e, st) {
+      _logger.warning('Package API CDN update failed.', e, st);
+    }
+    // Wait for 1 minutes for sanity, before trying again.
+    await Future.delayed(Duration(minutes: 1));
+  }
+}
+
+@visibleForTesting
+Future<void> doUploadPackageApisToCdnBucket(
+  GlobalLockClaim claim,
+  Bucket bucket, {
+  Duration sleepDuration = const Duration(minutes: 2),
+  int concurrency = _defaultUploadPackageApisToCdnBucketConcurrency,
+}) async {
+  final firstClaimed = clock.now();
+
+  // The claim will be released after a day, another process may
+  // start to upload the API JSONs from scratch again.
+  final workUntil = clock.now().add(Duration(days: 1));
+
+  Future<void> update(String package, bool isVisible) async {
+    if (isVisible) {
+      await _uploadPackageApiToCdnBucket(package, bucket);
+    } else {
+      await _deletesPackageFromBucket(package, bucket);
+    }
+  }
+
+  // initial scan of packages
+  final pool = Pool(concurrency);
+  final futures = <Future>[];
+  await for (final mp in dbService.query<ModeratedPackage>().run()) {
+    if (!claim.valid) {
+      return;
+    }
+    final f = pool.withResource(() => update(mp.name!, false));
+    futures.add(f);
+  }
+  await Future.wait(futures);
+  futures.clear();
+  if (!claim.valid) {
+    return;
+  }
+
+  await for (final package in dbService.query<Package>().run()) {
+    if (!claim.valid) {
+      return;
+    }
+    final f = pool.withResource(() => update(package.name!, package.isVisible));
+    futures.add(f);
+  }
+  await Future.wait(futures);
+  futures.clear();
+  if (!claim.valid) {
+    return;
+  }
+
+  // start monitoring
+  var lastQueryStarted = firstClaimed;
+  while (claim.valid) {
+    final now = clock.now().toUtc();
+    if (now.isAfter(workUntil)) {
+      break;
+    }
+
+    lastQueryStarted = now;
+
+    // query updates
+    final recentlyUpdated = await _queryRecentlyUpdated(lastQueryStarted);
+    for (final e in recentlyUpdated.entries) {
+      if (!claim.valid) {
+        break;
+      }
+      final f = pool.withResource(() => update(e.key, e.value));
+      futures.add(f);
+    }
+    await Future.wait(futures);
+    futures.clear();
+    await Future.delayed(sleepDuration);
+  }
+  await pool.close();
+}
+
+/// Uploads the package version API response bytes to the bucket, mirroring
+/// the endpoint name in the file location.
+Future<void> _uploadPackageApiToCdnBucket(
+  String package,
+  Bucket bucket,
+) async {
+  final data = await retry(() => packageBackend.listVersions(package));
+  final rawBytes = jsonUtf8Encoder.convert(data.toJson());
+  final gzippedBytes = gzip.encode(rawBytes);
+
+  for (final objectName in _cdnObjectNames(package)) {
+    final info = await bucket.tryInfo(objectName);
+    // Skip upload if was updated recently, and the bytes length and md5 hash matches.
```

Comment:
Why do we reupload after 12 hours? If the hash matches we should still be good - no?

Reply:
Maybe we don't need to when hashes match, but I'd rather be wrong by writing it more than we needed.